### PR TITLE
736 Add mimalloc as opt-in replacement memory allocator

### LINKED ISSUE #736: Add debugging memory allocator
**What Needs to be Done?**

Swap out glibc for another allocator to track usage better.

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #736

CMake options to enable:

```cmake
-Dvt_mimalloc_enabled:BOOL=1
-Dvt_mimalloc_static:BOOL=ON
-DMI_INTERPOSE:BOOL=ON
-DMI_OVERRIDE:BOOL=ON
```

Environment options to add stats at the end of run

```
export MIMALLOC_SHOW_STATS=1
```

Verbose mode:

```
export MIMALLOC_VERBOSE=1
```

### CODE REVIEW

**Comment by PhilMiller** [comment]
Does your import modify the upstream mimalloc code at all, or is it 100% stock?

**Comment by PhilMiller** [comment]
Ugh, this actually implements its own full allocator. That's way more intrusive than just being able to get instrumentation. It's not clear without digging into the docs - does it have a mode that just tracks (de)allocations, and can provide statistics, without overriding use of whatever would otherwise be called?

**Comment by lifflander** [comment]
@PhilMiller Yes, it implements it's own allocator. I couldn't find a instrumenting one that is correctly delegates to the original. I tried many variants. 

I just pushed the changes to mimalloc, which adds the incremental statistic.

File: lib/mimalloc/include/mimalloc.h [hunk 1]
```diff
@@ -359,6 +359,8 @@ mi_decl_nodiscard mi_decl_export mi_decl_restrict void* mi_new_n(size_t count, s
 mi_decl_nodiscard mi_decl_export void* mi_new_realloc(void* p, size_t newsize)                mi_attr_alloc_size(2);
 mi_decl_nodiscard mi_decl_export void* mi_new_reallocn(void* p, size_t newcount, size_t size) mi_attr_alloc_size2(2, 3);
 
+mi_decl_export size_t getAllocatedSize();
```
Comment (PhilMiller) [inline lib/mimalloc/include/mimalloc.h]: Matching style and such, this should probably be named `mi_get_allocated_size()`

Comment (lifflander) [reply lib/mimalloc/include/mimalloc.h]: Good point

File: CMakeLists.txt [hunk 2]
```diff
@@ -61,6 +61,18 @@ endif()
 add_subdirectory(${PROJECT_LIB_DIR}/CLI)
 add_subdirectory(${PROJECT_LIB_DIR}/fmt)
 
+option(vt_mimalloc_enabled "Build VT with mimalloc" OFF)
+option(vt_mimalloc_static "Build VT with mimalloc using static linking" ON)
+if (${vt_mimalloc_enabled})
```
Comment (pnstickne) [inline CMakeLists.txt]: `if (vt_mimalloc_enabled)` - cmake variable should be used directly in `if` unless the explicit expansion is desired.

File: cmake_config.h.in [hunk 3]
```diff
@@ -67,6 +67,7 @@
 #define vt_feature_cmake_priorities          @vt_feature_cmake_priorities@
 #define vt_feature_cmake_fcontext            @vt_feature_cmake_fcontext@
 #define vt_feature_cmake_test_trace_on       @vt_feature_cmake_test_trace_on@
+#define vt_feature_cmake_mimalloc            @vt_feature_cmake_mimalloc@
```
Comment (pnstickne) [inline cmake_config.h.in]: The can probably all be #cmakedefine

**Comment by pnstickne** [comment]
Would recommend adding a 'fetch' script which also prunes all non-necessary files, such as docs, tests, examples in the TPL to reduce the additions to that which is required for inclusion in VT's repo.

**Comment by pnstickne** [comment]
It might also be useful to collect the changes to mimalloc in this work into a patch also included somewhere in VT for when/if mimalloc is ever updated.

File: src/vt/utils/memory/memory_usage.cc [hunk 4]
```diff
@@ -267,6 +271,20 @@ std::string StatM::getName() {
   return "selfstatm";
 }
 
+std::size_t Mimalloc::getUsage() {
+# if backend_check_enabled(mimalloc)
```
Comment (pnstickne) [inline src/vt/utils/memory/memory_usage.cc]: What `#` style to follow?

File: lib/mimalloc/ide/vs2017/mimalloc-override-test.vcxproj [hunk 5]
```diff
@@ -0,0 +1,190 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (pnstickne) [inline lib/mimalloc/ide/vs2017/mimalloc-override-test.vcxproj]: Mimalloc development/support files can all be omitted from lib/

File: lib/mimalloc/.gitattributes [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# default behavior is to always use unix style line endings
```
Comment (pnstickne) [inline lib/mimalloc/.gitattributes]: Both .git files can be removed.

File: src/vt/utils/memory/memory_usage.cc (hunk 4 above)
Comment (lifflander) [reply src/vt/utils/memory/memory_usage.cc]: This style should be following. `#` starts on the very first character, with subsequent spaces.

File: lib/mimalloc/ide/vs2017/mimalloc-override-test.vcxproj (hunk 5 above)
Comment (lifflander) [reply lib/mimalloc/ide/vs2017/mimalloc-override-test.vcxproj]: Good point

File: lib/mimalloc/.gitattributes (hunk 6 above)
Comment (lifflander) [reply lib/mimalloc/.gitattributes]: Addressed

**Pull request merged**